File: Data Structures & Algorithms/pacific-atlantic-water-flow/submission-0.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        res = []
        pacific = set()
        atlantic = set()
        rows = len(heights)
        cols = len(heights[0])

        def dfs(r,c,h,o):
            if (r < 0 or r >= rows or
                c < 0 or c >= cols or 
                heights[r][c] < h or 
                (r,c) in o):
                return

            o.add((r,c))
            dfs(r, c+1, heights[r][c], o)
            dfs(r, c-1, heights[r][c], o)
            dfs(r+1, c, heights[r][c], o)
            dfs(r-1, c, heights[r][c], o)


        for c in range(cols):
            dfs(0,c,heights[0][c], pacific)
            dfs(rows-1, c, heights[rows-1][c], atlantic)

        
        for r in range(rows):
            dfs(r,0,heights[r][0], pacific)
            dfs(r, cols-1, heights[r][cols-1], atlantic)

    
        for r in range(rows):
            for c in range(cols):
                if (r,c) in pacific and (r,c) in atlantic:
                    res.append([r,c])

        return res

File: Data Structures & Algorithms/surrounded-regions/submission-0.cpp
class Solution {
public:
    void dfs(vector<vector<char>>& board,int i, int j) {
        // if you come across in the dfs an edge piece return false
        if (i<0 || i>=board.size() || j<0 || j>=board[0].size() || board[i][j] != 'O') return;
        board[i][j] = 'T';
        dfs(board,i+1,j);
        dfs(board,i-1,j);
        dfs(board,i,j+1);
        dfs(board,i,j-1);
    }

    void solve(vector<vector<char>>& board) {
        int ROWS = board.size();
        int COLS = board[0].size();
        for (int i=0; i<ROWS; i++) {
            for (int j=0; j<COLS; j++) {
                if (board[i][j] == 'O' && 
                ((i==0 || i==ROWS-1) ||
                (j==0 || j==COLS-1))) dfs(board,i,j);
            }
        }

        for (int i=0; i<ROWS; i++) {
            for (int j=0; j<COLS; j++) {
                if (board[i][j] == 'O') board[i][j] = 'X';
            }
        }
        
        for (int i=0; i<ROWS; i++) {
            for (int j=0; j<COLS; j++) {
                if (board[i][j] == 'T') board[i][j] = 'O';
            }
        }
    }
};
